### DISCUSSION TITLE
Roadmap (old)

### DISCUSSION THREAD

**Discussion by RoboMWM** [GriefPrevention/GriefPrevention#1162] [General]
- v17: remove extraneous features, refactor packages (as much as possible). Breakages: addons (API) (and the obviously removed features). May also be able to squeeze in some (potentially experimental) designs in here e.g. bounding box for claims.
    - Maintainability. Make plugin more maintainable/easier to maintain by software developers; make it easier to make fixes and add features.
- v18: finish refactoring packages (with exception of datastore), split config into a file for each primary feature, refactor config handling. Conversion code to bring v17 users over. Breakages: none expected
    - Configurability. Make plugin easier for server administrators to customize/disable features according to their needs.
- v19: refactor/near-rewrite of datastore/storage, implement event funnel. Breakages: those using MySQL dbs.
    - Reliability. Make plugin store and manage data more reliably.
    - Augmentable (cuz "augmentability" isn't a word?). Make developing addons easier to perform.
    - Maintainability. Make plugin easier to add fixes.
- v20: little too far to project, but mostly "cleanup" along with conversion code to transition from v16. Add any missing API.
    - Stability. Finalize plugin changes and standards so that changes don't break existing addons
    - Migratable(?). Provide migration path for those on v16.

may swap v18 and v19, if it's determined that the refactoring in v18 can't proceed without also performing significant changes to database-related classes.

**Comment by TomLewis** [comment]
Using Mysql database: Oof.

When you get around to doing the storage refactor, ping me and I can send you a copy of my live working database for testing!

**Reply by RoboMWM** [reply]
Ya, hence the major version bump. Very likely we'll keep GP v16 in an "extended support" phase where no active development will occur but bugfixes will be made until we hit v20.

Idea behind DataStore refactor other than cleaning out the excessively large abstract class is so other addons can easily implement the methods, thus allowing for any storage implementations desired.

**Reply by TomLewis** [reply]
Dec 2020, Bit of a late reply, but better late than never! As long as we get to keep MySQL and theres a conversion SQL pre-written if the database schema changes then it wont break anyones servers!

**Comment by RoboMWM** [comment]
Since there will be some interesting world changes with Minecraft 1.17, I'll likely be holding off until then before beginning development on GP v17. I think that this will be best so that the majority of current servers can enjoy a relatively stable version of GP for the Minecraft 1.17 without the community being bogged down/confused with effectively two versions of GP in development.

Idea is v16 will be maintained in some capacity until v20 is ready to go. Once the branch for v17 occurs, PRs that perform significant refactoring or add major features in v16 are likely to be ignored as all priority will be on v17 onwards. I'm not sure what the term for this is... is it "extended support?" Anyway, this means v17, 18, 19 will all be "beta" quality releases in terms of feature completeness/data migration. This will also allow for these versions to have more change to underlying structures without worrying too much about the pressure a regular release requires (in terms of testing, reviewing, "informal QA", etc.)

**Reply by TomLewis** [reply]
Current LTS, Ending LTS support into "End of Life" is the release categories for Ubuntu, you could use these terms to explain releases.
https://wiki.ubuntu.com/Releases

- Current
- Future
- Extended Security Maintenance
- End of Life


It might be worth making a table on the Wiki or README.md that explains all versions, support & future planning in one go, including Minecraft versions and the version of GriefPrevention people need to download.
Then it would be all in one nice table!

If you really wanted to get fancy, Ubuntu have a release cycle page that explains it all in a graphic too that you could get inspiration from! https://ubuntu.com/about/release-cycle

**Reply by RoboMWM** [reply]
Cool, thanks for the tip! I think Windows uses something similar as well.

**Comment by RoboMWM** [comment]
Gradual changes are gonna be the best way instead of a hard break. I may still do a hard break when certain features get removed (siege, etc.).

On the topic of removing features; should try a "phase out" or deprecation-esque approach:

First step is to migrate/create documentation of all features we wish to retain.

Then the next step is to remove features we want to eventually cut out from being added to default configuration.

**Reply by TomLewis** [reply]
Ive always thought fitlering chat for bad words was not in the scope at all with GriefPrevention, Most people use a chat plugin to manage thier chat, that can definitally go, infact the entire "spam" section can be replaced with nearly any chat plugin.

The PvP section may also be replaced by PvPManager, you could maybe hook into PvPManager to support specific features, like Siege.

Of all the features to go, Siege is the last thing I want to loose! Hopefully this will be remade at the same time https://github.com/GPAddons/GPSiege as there is currently a bug that no siege soft blocks are working! https://github.com/TechFortress/GriefPrevention/issues/1794
